libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+    for (const uint8_t *ptr = buffer; ptr < b; ++ptr) {
+        crc = crc8_dvb_s2(crc, *ptr);
+    }
+
+    *b = crc;
+
+    // send data if possible
+    uart->write(buffer, 3 + (have_param ? 1 : 0));
+}
+
+// crc functions
+uint8_t AP_RunCam::crc8_dvb_s2(uint8_t crc, uint8_t a)
+{
+    crc ^= a;
+    for (uint8_t i = 0; i < 8; ++i) {
+        if (crc & 0x80) {
+            crc = (crc << 1) ^ 0xD5;
+        } else {
+            crc = crc << 1;
+        }
+    }
+    return crc;
+}
+
+uint8_t AP_RunCam::crc8_high_first(uint8_t *ptr, uint8_t len)
+{
+    uint8_t crc = 0x00;
+    while (len--) {
+        crc ^= *ptr++;
+        for (uint8_t i = 8; i > 0; --i) {
+            if (crc & 0x80)
+                crc = (crc << 1) ^ 0x31;
+            else
+                crc = (crc << 1);
+        }
+    }
+    return (crc);
+}
+
+// handle a device info response
+void AP_RunCam::parse_device_info(const Response& response)
+{
+    _protocol_version = ProtocolVersion(response._recv_buf[1]);
+
+    uint8_t featureLowBits = response._recv_buf[2];
+    uint8_t featureHighBits = response._recv_buf[3];
+    if (_features == 0) {
+        _features = (featureHighBits << 8) | featureLowBits;
+    }
+    _state = State::INITIALIZED;
+    gcs().send_text(MAV_SEVERITY_INFO, "RunCam device initialized, features 0x%04X\n", _features.get());
+}
+
+// wait for the RunCam device to be fully ready
+bool AP_RunCam::camera_ready()
+{
+    if (_state != State::INITIALIZING && _state != State::INITIALIZED) {
+        return true;
+    }
+    return false;
+}
+
+// error handler for OSD simulation
+void AP_RunCam::simulation_OSD_cable_failed(const Response& response)
+{
+    _waiting_device_response = false;
+    if (response._command == Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION) {
+        uint8_t operationID = response._param;
+        if (operationID == uint8_t(ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_CLOSE)) {
+            return;
+        }
+    }
+}
+
+// constructor for a response structure
+AP_RunCam::Response::Response(AP_RunCam* device, Command commandID, uint8_t param,
+    uint32_t timeout, uint16_t maxRetryTimes, parse_func_t parserFunc)
+    : _recv_buf(device->_recv_buf),
+    _command(commandID),
+    _max_retry_times(maxRetryTimes),
+    _timeout_ms(timeout),
+    _device(device),
+    _param(param),
+    _parser_func(parserFunc)
+{
+    _request_timestamp_ms = AP_HAL::millis();
+    _expected_response_length = get_response_length(commandID);
+}
+
+// get the length of a response
+uint8_t AP_RunCam::Response::get_response_length(const Command command)
+{
+    for (uint16_t i = 0; i < RUNCAM_NUM_EXPECTED_RESPONSES; i++) {
+        if (_expected_responses_length[i].command == command) {
+            return _expected_responses_length[i].reponse_length;
+        }
+    }
+
+    return 0;
+}
+
+// process all of the pending responses, retrying as necessary
+bool AP_RunCam::process_waiting(Response& response, uint32_t now)
```

Comment:
`Response` is really `Request`.  It's a request that's waiting for a response.

Reply:
Gah! Ok you are right